src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -111685,5 +111685,184 @@ void M(object? o1)
                     Diagnostic(ErrorCode.WRN_ConvertingNullableToNonNullable, "o1").WithLocation(8, 21));
             Assert.True(comp.ShouldRunNullableWalker);
         }
+
+        [Theory]
+        // 2 states * 3 states * 3 states = 18 combinations
+        [InlineData("locationNonNull", "valueNonNull", "comparandNonNull", true)]
+        [InlineData("locationNonNull", "valueNonNull", "comparandMaybeNull", true)]
+        [InlineData("locationNonNull", "valueNonNull", "null", true)]
+        [InlineData("locationNonNull", "valueMaybeNull", "comparandNonNull", false)]
+        [InlineData("locationNonNull", "valueMaybeNull", "comparandMaybeNull", false)]
+        [InlineData("locationNonNull", "valueMaybeNull", "null", false)]
+        [InlineData("locationNonNull", "null", "comparandNonNull", false)]
+        [InlineData("locationNonNull", "null", "comparandMaybeNull", false)]
+        [InlineData("locationNonNull", "null", "null", false)]
+        [InlineData("locationMaybeNull", "valueNonNull", "comparandNonNull", false)]
+        [InlineData("locationMaybeNull", "valueNonNull", "comparandMaybeNull", false)]
+        [InlineData("locationMaybeNull", "valueNonNull", "null", true)]
+        [InlineData("locationMaybeNull", "valueMaybeNull", "comparandNonNull", false)]
+        [InlineData("locationMaybeNull", "valueMaybeNull", "comparandMaybeNull", false)]
+        [InlineData("locationMaybeNull", "valueMaybeNull", "null", false)]
+        [InlineData("locationMaybeNull", "null", "comparandNonNull", false)]
+        [InlineData("locationMaybeNull", "null", "comparandMaybeNull", false)]
+        [InlineData("locationMaybeNull", "null", "null", false)]
+        [WorkItem(36911, "https://github.com/dotnet/roslyn/issues/36911")]
+        public void CompareExchange_LocationNullState(string location, string value, string comparand, bool isLocationNonNull)
+        {
+            var source = $@"
+#pragma warning disable 0436
+using System.Threading;
+namespace System.Threading
+{{
+	public static class Interlocked
+    {{
+        public static object? CompareExchange(ref object? location, object? value, object? comparand) {{ return location; }}
+    }}
+}}
+
+class C
+{{
+    void M(object? locationNonNull, object? locationMaybeNull, object comparandNonNull, object? comparandMaybeNull, object valueNonNull, object? valueMaybeNull)
+    {{
+        locationNonNull = new object();
+        Interlocked.CompareExchange(ref {location}, {value}, {comparand});
```

Comment:
Consider testing the return of CompareExchange as well.

Reply:
Sure, I think demonstrating the current behavior is a good idea. Will create a tracking issue for analysis of the return values.